cpp/src/unary/cast_ops.cu

```diff
@@ -0,0 +1,245 @@
+ */
+
+#include "unary_ops.cuh"
+
+#include "rmm/thrust_rmm_allocator.h"
+#include "utilities/wrapper_types.hpp"
+#include "utilities/type_dispatcher.hpp"
+
+#include <thrust/copy.h>
+#include <thrust/execution_policy.h>
+
+
+template<typename TypeFrom, typename TypeTo>
+struct DeviceCast {
+    __device__
+    TypeTo apply(TypeFrom data) {
+        return static_cast<TypeTo>(cudf::detail::unwrap(data));
+    }
+};
+
+template<typename TypeFrom, typename TypeTo, int64_t units_factor>
+struct UpCasting {
+    __device__
+    TypeTo apply(TypeFrom data) {
+        auto udata = cudf::detail::unwrap(data);
+        return static_cast<TypeTo>(udata*units_factor);
+    }
+};
+
+template<typename TypeFrom, typename TypeTo, int64_t units_factor>
+struct DownCasting {
+    __device__
+    TypeTo apply(TypeFrom data) {
+        auto udata = cudf::detail::unwrap(data);
+        return static_cast<TypeTo>((udata-(units_factor-1)*(udata<0))/units_factor); //ceiling only when data is negative
+    }
+};
+
+
+template <typename TypeFrom>
+struct CastTo_Dispatcher {
+    template <typename TypeTo>
+    gdf_error operator()(gdf_column *input, gdf_column *output) {
+        return cudf::unary::Launcher< TypeFrom, TypeTo, DeviceCast<TypeFrom, TypeTo> >::launch(input, output);
+    }
+};
+
+
+template <typename TypeFrom> // This can be either cudf::date32 or cudf::date64
+struct CastDateTo_Dispatcher {
+    // Cast to other date type
+    template <typename TypeTo>
+    typename std::enable_if_t<
+        std::is_same<TypeTo, cudf::date32>::value ||
+        std::is_same<TypeTo, cudf::date64>::value,
+    gdf_error>
+    operator()(gdf_column *input, gdf_column *output) {
+        if( input->dtype == GDF_DATE64 && output->dtype == GDF_DATE32 ) 
+            return cudf::unary::Launcher<TypeFrom, TypeTo, DownCasting<TypeFrom, TypeTo, 86400000> >::launch(input, output);
+        else if( input->dtype == GDF_DATE32 && output->dtype == GDF_DATE64 ) 
+            return cudf::unary::Launcher<TypeFrom, TypeTo, UpCasting<TypeFrom, TypeTo, 86400000> >::launch(input, output);
+        else // both columns have same type
+            return cudf::unary::Launcher< TypeFrom, TypeTo, DeviceCast<TypeFrom, TypeTo> >::launch(input, output);
+    }
+    
+    // Cast to cudf::timestamp
+    template <typename TypeTo>
+    typename std::enable_if_t<
+        std::is_same<TypeTo, cudf::timestamp>::value,
+    gdf_error>
+    operator()(gdf_column *input, gdf_column *output) {
+        if( input->dtype == GDF_DATE32 && ( output->dtype == GDF_TIMESTAMP && output->dtype_info.time_unit == TIME_UNIT_s ) ) 
+            return cudf::unary::Launcher<TypeFrom, TypeTo, UpCasting<TypeFrom, TypeTo, 86400> >::launch(input, output);
+        else if( input->dtype == GDF_DATE32 && ( output->dtype == GDF_TIMESTAMP && output->dtype_info.time_unit == TIME_UNIT_ms ) ) 
+            return cudf::unary::Launcher<TypeFrom, TypeTo, UpCasting<TypeFrom, TypeTo, 86400000> >::launch(input, output);
```

Comment:
What are these magic numbers? 86400, 86400000, etc?

Reply:
I see they are units factors. Suggest defining them all together in a clear and documented place as `constexpr`s with clear names.